### ISSUE TITLE
CNAME to third-party domain in FPS

### ISSUE THREAD

**Issue by RussStringham** [privacycg/first-party-sets#79]
**Summary**
A site may want to list in its FPS a subdomain that maps via a CNAME to a third-party domain. In doing so, they would be claiming that any data going to the third-party domain is under their control and their privacy policy also applies to this data. Browsers and privacy plugins may then choose to relax restrictions that some have put in place for third-party CNAMEs.

Technically, this is not an FPS issue, because the CNAME is already a part of the first-party, but since some browser and privacy plugins are treating it like a third-party, supporting it in FPS allows it to be brought back into the first-party context.

**Background**
A CNAME is an alias that maps a (sub)domain to another (sub)domain. When a link references the CNAME domain, the mapping is followed so that the resource is loaded from the target domain. The target can be another CNAME, in which case it is also followed until a final target is reached. If the source and final target are not part of the same domain, then this is referred to as a third-party CNAME.

Use of a third-party CNAME is common when a site wants to contract for a service offered by a different company for which they don't have the time or expertise to implement themselves, or because they would prefer to focus their energy on their core products and services. Common uses of a third-party CNAME include to map to a CDN and to provide tech/product support services via a third-party within the site itself. As another use case, most websites do not want to implement their own web analytics, but instead use a third-party's services to provide this functionality. Because cookies and sometimes even the calls to these third-party services are blocked by some browsers and plugins, the site may wish to enable first-party data collection by mapping the third-party site to a CNAME in their own domain. Most Adobe Analytics customers use a CNAME for this purpose, because they trust Adobe with their data (I work on the Adobe Analytics product). For example, a site such somecompany.example will configure a CNAME for metrics.somecompany.example that maps to omtrdc.net (Adobe's data collection domain). A cookie created by omtrdc.net can then be stored under the somecompany.example domain.

Using a third-party CNAME requires a significant amount of trust by the first-party, because the mapped site is traditionally granted full access to all site cookies and functions within the security context of the primary domain. In Adobe's case, the site owner who has contracted with Adobe maintains ownership of the data and Adobe follows their directions in how the data is used. In GDPR terms, the site owner is the _data controller_ while Adobe is only a _data processor_.

Recent changes in Safari break this workflow because they treat a third-party CNAME as if it were a third-party site (some privacy plugins have done this for longer). However, Apple itself uses a CNAME on its own sites to enable first-party data collection to Adobe Analytics. When the first-party site owns and controls the data, it should be allowed to utilize services that improve its offerings.

While Adobe's omtrdc.net domain is used to collect data for many different sites that have mapped it using a CNAME to their own domain, I am not aware of any way that Adobe could even use omtrdc.net to track a user between these sites, because each CNAME reference to omtrdc.net has its own cookies and own local storage; nothing is shared except the destination where the data is sent. I suppose that fingerprinting could be utilized to tie together data coming from the same user via different CNAMEs, but fingerprinting works just as effectively without the CNAME and should be disallowed anyway for all cross-site tracking.

Note that sites can work around all browser implemented third-party CNAME restrictions by changing the CNAME entry to refer to a proxy server in the first party domain, where the proxy forwards all calls to the third-party and returns its responses to the client. However, this adds complexity, introduces additional points of failure, increases the site's costs, increases response times and harms the end-user experience.

**Ask**
Allow a third-party CNAME to be included within the definition of an FPS, but only if:

- the first-party maintains control of the data.
- the first-party's privacy policy applies to all data exchanged via the CNAMEd service.
- fingerprinting or similar techniques will not be utilized by the third-party site accessed via the CNAME to enable cross-site tracking.

When the third-party CNAME is defined in the FPS, the user agent may elect to reduce restrictions that otherwise apply.

**Comment by krgovind** [comment]
@RussStringham Thank you for the detailed description!

Based on your example usecase where `metrics.somecompany.example` is CNAME'd to `omtrdc.net`; is the ask that it be allowed to include `metrics.somecompany.example` in the website's First-Party Set? Since `metrics.somecompany.example` is a subdomain of `somecompany.example`; I don't think a FPS is needed at all; since most browsers already consider all subdomains as being the same "first-party"?

Regarding the Safari change to third-party CNAMEs, is this the mitigation you are referring to: [CNAME Cloaking and Bounce Tracking Defense](https://webkit.org/blog/11338/cname-cloaking-and-bounce-tracking-defense/)? If so, I'm not sure what threat the mitigation is protecting against (since CNAME'd domains' cookies are equivalent to partitioned cookies in my mind), so we can discuss this on Thursday's call to ensure alignment with Safari's protections.

Speaking as a member of the Chrome team, it seems like [partitioned cookies](https://github.com/WICG/CHIPS) is the right solution for this use-case, instead of CNAME aliasing.

**Comment by RussStringham** [comment]
> Since metrics.somecompany.example is a subdomain of somecompany.example; I don't think a FPS is needed at all; since most browsers already consider all subdomains as being the same "first-party"?

Agree that FPS shouldn't be needed in this case, but Safari isn't treating the CNAME target as equivalent to a first-party subdomain and there is concern that Firefox and/or Edge might follow suit if the third-party domain is on their disconnect.me tracker list. FPS seems like it could be a place for a site to declare that this usage is not being used for cross-site tracking, but is simply part of the FPS (i.e. the third-party site is part of the FPS, but only when accessed via the CNAME). 

Partitioned cookies seem like a good alternative, but in this case they are not really needed and don't help, because the cookie is already owned by the first-party domain.

**Comment by krgovind** [comment]
> Agree that FPS shouldn't be needed in this case, but Safari isn't treating the CNAME target as equivalent to a first-party subdomain and there is concern that Firefox and/or Edge might follow suit if the third-party domain is on their disconnect.me tracker list. FPS seems like it could be a place for a site to declare that this usage is not being used for cross-site tracking, but is simply part of the FPS (i.e. the third-party site is part of the FPS, but only when accessed via the CNAME).

Since I can't speak for Safari, Firefox, and Edge's potential usage of First-Party Sets; I will propose this discussion as an agenda item for this Thursday's PrivacyCG call. If you are able to attend, please do so.

> Partitioned cookies seem like a good alternative, but in this case they are not really needed and don't help, because the cookie is already owned by the first-party domain.

Right. I was suggesting whether perhaps an alternative way of achieving the same effective semantics might be to not use the CNAME alias at all; but to have `omtrdc.net` use partitioned cookies instead and have `somecompany.example` directly embed resources from `omtrdc.net`. I think this would conform to the various browsers' privacy models, and won't run into the issues of CNAME related privacy protections in browsers.

**Comment by krgovind** [comment]
@RussStringham - My apologies for the late follow-up. Based on the Feb-10 [meeting discussion](https://github.com/privacycg/meetings/blob/main/2022/telcons/02-10-minutes.md#cname-to-third-party-domain-in-fps), members of the PrivacyCG were concerned about the security implications of CNAMEs, and did not want to encourage their use. Instead, use of partitioned cookies is likely the better path here. I'd recommend that you watch for upcoming work on [CHIPS](https://github.com/WICG/CHIPS), which is being adopted as a PrivacyCG work item. Another existing PrivacyCG work item that may be of interest to you is [Client-side storage partitioning](https://github.com/privacycg/storage-partitioning) which applies to storage other than cookies (indexed DB, localStorage, etc.)

The equivalent of Safari's [CNAME Cloaking and Bounce Tracking Defense](https://webkit.org/blog/11338/cname-cloaking-and-bounce-tracking-defense/) would be to have a similar 7-day cap on partitioned cookies and storage. 

As currently defined, CHIPS and (from what I can tell) storage partitioning do not have a similar cap; but it is possible that different browsers will apply different policies. It might be best to provide feedback on those proposals to discuss how a lifetime cap might impact your usecase.

I am closing this issue now; but please feel free to re-open in the future if leveraging FPS seems like a promising path forward.
